pkgs/development/python-modules/discordpy/default.nix

```diff
@@ -6,62 +6,62 @@
   fetchFromGitHub,
   libopus,
   pynacl,
-  pythonOlder,
   withVoice ? true,
   ffmpeg,
+  setuptools,
 }:
 
-buildPythonPackage rec {
+let
   pname = "discord.py";
   version = "2.4.0";
-  format = "setuptools";
-
-  disabled = pythonOlder "3.8";
+in
+buildPythonPackage {
+  inherit pname version;
+  pyproject = true;
 
   src = fetchFromGitHub {
     owner = "Rapptz";
-    repo = pname;
+    repo = "discord.py";
     rev = "refs/tags/v${version}";
     hash = "sha256-GIwXx7bRCH2+G3zlilJ/Tb8el50SDbxGGX2/1bqL3+U=";
   };
 
-  propagatedBuildInputs =
+  build-system = [ setuptools ];
+
+  dependencies =
     [ aiohttp ]
     ++ lib.optionals withVoice [
+      ffmpeg
       libopus
       pynacl
-      ffmpeg
     ];
 
-  patchPhase =
-    ''
-      substituteInPlace "discord/opus.py" \
-        --replace "ctypes.util.find_library('opus')" "'${libopus}/lib/libopus${stdenv.hostPlatform.extensions.sharedLibrary}'"
-    ''
-    + lib.optionalString withVoice ''
-      substituteInPlace "discord/player.py" \
-        --replace "executable='ffmpeg'" "executable='${ffmpeg}/bin/ffmpeg'"
-    '';
+  patchPhase = lib.optionalString withVoice ''
+    substituteInPlace "discord/opus.py" \
+      --replace-fail "ctypes.util.find_library('opus')" "'${libopus}/lib/libopus${stdenv.hostPlatform.extensions.sharedLibrary}'"
+
+    substituteInPlace "discord/player.py" \
+      --replace-fail "executable: str = 'ffmpeg'" "executable: str = '${lib.getExe ffmpeg}'"
+  '';
```

Comment:
Loosely related, do we even need this? Can the lib not auto detect the `ffmpeg` binary without doing this replace? Also note that this replace was broken since [`discord.py` version `2.0.0`](https://github.com/Rapptz/discord.py/releases/tag/v2.0.0)

Reply:
it'll work if ffmpeg is available in PATH. but propagating things that go in `bin` is not common

Reply:
So, as long as we have `ffmpeg` as a runtime dependency, we don't need to do this substitution, right?

Reply:
I tested a `nix-shell`, and ffmpeg is indeed added to PATH. go ahead, but please move `ffmpeg` from `dependencies` and over to  `propagatedBuildInputs`, since it should not be added to the python `passthru.dependencies`

Reply:
The same should be done for `libopus`?

On a side note, I've asked upstream why they need `libopus` when `python3.opuslib` exists.

Reply:
I'm done thinking out loud now, I think i still prefer patching over propagation

Reply:
Ah, since we're adding `ffmpeg` and `libopus` in the patch anyway, we don't really need them in `dependencies` _or_ `propagatedBuildInputs`, right?

Reply:
right